flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -900,33 +901,42 @@ template <typename OP>
 struct EmboxCommonConversion : public FIROpConversion<OP> {
   using FIROpConversion<OP>::FIROpConversion;
 
+  /// Type of lambda helper to abstract whether the embox is lowered as StoreOps
+  /// or InsertOps.
+  using StoreOrInserter = std::function<mlir::Value(
+      mlir::ConversionPatternRewriter &, mlir::Value, llvm::ArrayRef<unsigned>,
+      mlir::Value,
+      const std::function<mlir::Value(mlir::LLVM::LLVMType, mlir::Value)> &)>;
+
   /// Generate an alloca of size `size` and cast it to type `toTy`
   mlir::LLVM::AllocaOp
   genAllocaWithType(mlir::Location loc, mlir::LLVM::LLVMType toTy,
                     unsigned alignment,
                     mlir::ConversionPatternRewriter &rewriter) const {
     auto thisPt = rewriter.saveInsertionPoint();
     auto *thisBlock = rewriter.getInsertionBlock();
-    auto func = mlir::cast<mlir::LLVM::LLVMFuncOp>(thisBlock->getParentOp());
-    rewriter.setInsertionPointToStart(&func.front());
+    if (auto func =
+            mlir::dyn_cast<mlir::LLVM::LLVMFuncOp>(thisBlock->getParentOp()))
+      rewriter.setInsertionPointToStart(&func.front());
     auto sz = this->genConstantOffset(loc, rewriter, 1);
     auto al = rewriter.create<mlir::LLVM::AllocaOp>(loc, toTy, sz, alignment);
     rewriter.restoreInsertionPoint(thisPt);
     return al;
   }
 
-  template <typename... FLDS>
-  mlir::LLVM::GEPOp genGEPToField(mlir::Location loc, mlir::LLVM::LLVMType ty,
-                                  mlir::ConversionPatternRewriter &rewriter,
-                                  mlir::Value base, mlir::Value zero,
-                                  FLDS... fields) const {
-    return this->genGEP(loc, ty.getPointerTo(), rewriter, base, zero,
-                        this->genConstantOffset(loc, rewriter, fields)...);
-  }
-
-  static mlir::LLVM::LLVMType getBoxEleTy(mlir::LLVM::LLVMType boxPtrTy,
-                                          unsigned i) {
-    return boxPtrTy.getPointerElementTy().getStructElementType(i);
+  // Get the element type given an LLVM type that is of the form
+  // [llvm.ptr](llvm.array|llvm.struct)+ and the provided indexes.
+  static mlir::LLVM::LLVMType getBoxEleTy(mlir::LLVM::LLVMType type,
+                                          llvm::ArrayRef<unsigned> indexes) {
+    if (type.isPointerTy())
+      type = type.getPointerElementTy();
+    for (auto i : indexes) {
+      if (type.isStructTy())
+        type = type.getStructElementType(i);
+      else if (type.isArrayTy())
+        type = type.getArrayElementType();
+    }
+    return type;
   }
 
   int getCFIAttr(fir::BoxType boxTy) const {
```

Comment:
This might be a bug on my part. We should check whether to follow a "callee side invariant" or a "called side adaptive" strategy for these bits. If the former, we should set the bit based on the memory reference *being* boxed and not how it will be used.

Reply:
Agreed about the need to investigate the requirement here. Right now this is called with the box type (and not the base address type of the embox), so it follows it follows the "callee side invariant" (how it will be used).